dlpy/embedding_model.py

```diff
@@ -0,0 +1,424 @@
+from dlpy.lr_scheduler import _LRScheduler, FixedLR, StepLR, FCMPLR
+
+
+class EmbeddingModel(Model):
+    input_layer_name_prefix = 'InputLayer_'
+    embedding_layer_name_prefix = 'EmbeddingLayer_'
+    embedding_loss_layer_name = 'EmbeddingLossLayer'
+    number_of_branches = 0
+    data_specs = None
+    embedding_model_type = None
+    branch_input_tensor = None
+    branch_output_tensor = None
+
+    @classmethod
+    def build_embedding_model(cls, branch, model_table=None, embedding_model_type='Siamese',
+                              embedding_layer=None, margin=None):
+        '''
+
+        Build an embedding model based on a given model branch and model type
+
+        Parameters
+        ----------
+        conn : CAS
+            Specifies the CAS connection object.
+        model_table : string or dict or CAS table, optional
+            Specifies the CAS table to store the deep learning model.
+            Default: None
+        branch : Model
+            Specifies the base model that is used as branches for embedding model.
+        embedding_model_type : string, optional
+            Specifies the embedding model type that the created table will be applied for training.
+            Valid values: Siamese, Triplet, and Quartet.
+            Default: Siamese
+        embedding_layer: Layer, optional
+            Specifies a dense layer as the embedding layer. For instance, Dense(n=10, act='identity') defines
+            the embedding dimension is 10. When it is not given, the last layer (except the task layers)
+            in the branch model will be used as the embedding layer.
+        margin: double, optional
+            Specifies the margin value used by the embedding model. When it is not given, for Siamese, margin is 2.0.
+            Otherwise, margin is 0.0.
+
+        Returns
+        -------
+        :class:`Model`
+
+        '''
+
+        # check the branch type
+        if not isinstance(branch, Model):
+            raise DLPyError('The branch option must contain a valid model')
+
+        # the branch must be built using functional APIs
+        # only functional model has the attr output_layers
+        if not hasattr(branch, 'output_layers'):
+            print("NOTE: Convert the branch model into a functional model.")
+            branch_tensor = branch.to_functional_model()
+        else:
+            branch_tensor = deepcopy(branch)
+
+        # always reset this local tensor to 0
+        branch_tensor.number_of_instances = 0
+
+        # the branch cannot contain other task layers
+        if len(branch_tensor.output_layers) != 1:
+            raise DLPyError('The branch model cannot contain more than one output layer')
+        elif branch_tensor.output_layers[0].type == 'output':
+            print("NOTE: Remove the output layers from the model.")
+            branch_tensor.layers.remove(branch_tensor.output_layers[0])
+            branch_tensor.output_layers[0] = branch_tensor.layers[-1]
```

Comment:
I suggest adding more types of task layer checking. For example, if there is a keypoint layer, it should be removed as well.

Reply:
@wenyushi helped change this. And will push later.